torch/distributed/_shard/checkpoint/state_dict_loader.py

```diff
@@ -131,26 +138,64 @@ def load_state_dict(
-        state_dict=state_dict, metadata_from_storage=metadata
-    )
-    bytes_futures = storage_reader.read_bytes(bytes_read_requests)
-    tensor_futures = storage_reader.read_tensors(tensor_read_requests)
-
-    bytes_futures.wait()
 
-    # Addtional steps are required to convert the bytes to its original type
-    # Note that this is NOT inplace,
-    # it creating a new object and replace what's in the state dict
-    for req in bytes_read_requests:
-        # Ensure the BytesIO is rewound
-        req.bytes.seek(0)
-        state_dict[req.fqn] = torch.load(req.bytes)
-
-    tensor_futures.wait()
+    .. note:: load_state_dict uses collectives to coordinate reads across ranks.
+        For NCCL-based process groups, internal tensor representations of objects
+        must be moved to the GPU device before communication takes place. In this
+        case, the device used is given by ``torch.cuda.current_device()`` and it
+        is the user's responsibility to ensure that this is set so that each rank
+        has an individual GPU, via ``torch.cuda.set_device()``
+    """
+    is_coordinator = not dist.is_initialized() or dist.get_rank(process_group) == coordinator_rank
+
+    try:
+        metadata = storage_reader.read_metadata()
+        bytes_read_requests, tensor_read_requests = _reshard_and_prepare_read_request(
+            state_dict=state_dict, metadata_from_storage=metadata
+        )
+        bytes_futures = storage_reader.read_bytes(bytes_read_requests)
+        tensor_futures = storage_reader.read_tensors(tensor_read_requests)
+
+        bytes_futures.wait()
+
+        # Addtional steps are required to convert the bytes to its original type
+        # Note that this is NOT inplace,
+        # it creating a new object and replace what's in the state dict
+        for req in bytes_read_requests:
+            # Ensure the BytesIO is rewound
+            req.bytes.seek(0)
+            state_dict[req.fqn] = torch.load(req.bytes)
+
+        tensor_futures.wait()
+        result = None
+    except BaseException as e:
+        result = e
+
+    global_result: List[Optional[CheckpointException]] = [None]
+    if dist.is_initialized():
+        # FIXME we could do an all_to_all_object if once existed
+        all_errors = [None] * dist.get_world_size(process_group)
+
+        dist.gather_object(
+            obj=result,
+            object_gather_list=all_errors if is_coordinator else None,
+            dst=coordinator_rank,
+        )
+
+        if is_coordinator:
+            node_failures = cast(Dict[int, BaseException], {i: err for i, err in enumerate(all_errors) if err is not None})
+            if len(node_failures) > 0:
+                global_result[0] = CheckpointException("failed to read checkpoint", node_failures)
+
+        dist.broadcast_object_list(
+            object_list=global_result,
+            group=process_group,
+            src=coordinator_rank)
```

Comment:
Can't we use `all_gather_object` instead of `gather` and then `broadcast`?

Reply:
Yeah, it required all ranks to locally aggregate errors, but I don't see any issue in doing it this way